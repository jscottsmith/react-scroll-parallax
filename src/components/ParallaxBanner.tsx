import React, { PropsWithChildren } from 'react';
import { Parallax } from './Parallax';

const containerStyle = {
  position: 'relative',
  overflow: 'hidden',
  width: '100%',
  height: '50vh',
};

const absoluteStyle = {
  position: 'absolute',
  top: 0,
  right: 0,
  bottom: 0,
  left: 0,
};

export interface BannerLayer {
  /**
   * A value from `-1` to `1` that represents the vertical offset to be applied to the current
   * layer, `0.1` would equal a `10%` offset on the top and bottom.
   */
  amount: number;
  /**
   * Custom layer children provided as a React element, for example `<Video />`
   */
  children?: any;
  /**
   * Indicate if the layer should be expanded with negative top/bottom margins so the edges will
   * never be visible.
   */
  expanded?: boolean;
  /**
   * Image source that will be applied as a CSS background image on the layer.
   */
  image?: string;
  /*
   * 	Props to apply to the layer element.
   */
  props?: any;
}

export interface ParallaxBannerProps {
  /**
   * Optionally pass additional class names to be added to the outermost parallax banner element.
   */
  className?: string;
  /**
   * Determines if the internal parallax layers will have offsets applied.
   */
  disabled?: boolean;
  /**
   * A required Array of Objects with layer properties: `[{ amount: 0.1, image: 'foo.jpg' }]`.
   */
  layers: BannerLayer[];
  /**
   * Optionally pass a style object to be added to the outermost parallax banner element.
   */
  style?: any;
}

const ParallaxBanner = ({
  children,
  className,
  layers,
  style,
  disabled,
}: PropsWithChildren<ParallaxBannerProps>) => {
  return (
    <div
      style={{ ...containerStyle, ...style }}
      className={'parallax-banner' + (className ? ` ${className}` : '')}
    >
      {layers.map(
        (
          {
            amount,
            children: layerChildren,
            expanded = true,
            image,
            props = {},
          },
          i
        ) => {
          // save props to be merged
          const layerStyle = props.style || {};
          const layerClass = props.className || '';

          // remove from pass through props
          delete props.style;
          delete props.className;

          const layerClassMerged = `parallax-banner-layer-${i}${
            layerClass ? ` ${layerClass}` : ''
          }`;

          // if this is an expanded layer overwrite the top/bottom styles with negative margins
          const expandedStyle = expanded
            ? {
                top: Math.abs(amount) * 100 * -1 + '%',
                bottom: Math.abs(amount) * 100 * -1 + '%',
              }
            : {};
          // optional image styles
          const imageStyle = image
            ? {
                backgroundImage: `url(${image})`,
                backgroundPosition: 'center',
                backgroundSize: 'cover',
              }
            : {};

          return (
            <Parallax
              key={`layer-${i}`}
              y={[amount * -1 * 100 + '%', amount * 100 + '%']}
              styleInner={absoluteStyle}
              styleOuter={absoluteStyle}
              disabled={disabled}
            >
              <div
                className={layerClassMerged}
                style={{
                  ...imageStyle,
                  ...absoluteStyle,
                  ...expandedStyle,
                  ...layerStyle,
                }}
                {...props}
              >
                {layerChildren}
              </div>
            </Parallax>
          );
        }
      )}
      {children}
    </div>
  );
};

ParallaxBanner.defaultProps = {
  disabled: false,
};

export default ParallaxBanner;
