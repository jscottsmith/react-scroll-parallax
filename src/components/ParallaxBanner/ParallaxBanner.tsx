import React, {
  PropsWithChildren,
  CSSProperties,
  useEffect,
  useRef,
  useState,
  ReactElement,
} from 'react';
import { ParallaxBannerLayer } from './components/ParallaxBannerLayer';
import { ParallaxBannerProps } from './types';

const containerStyle: CSSProperties = {
  position: 'relative',
  overflow: 'hidden',
  width: '100%',
};

export const ParallaxBanner = (
  props: PropsWithChildren<ParallaxBannerProps>
) => {
  const [targetElement, setTargetElement] =
    useState<HTMLDivElement | null>(null);
  const containerRef = useRef<HTMLDivElement>(null);
  useEffect(() => {
    setTargetElement(containerRef.current);
  }, []);
  const {
    disabled: disableAllLayers,
    style: rootStyle,
    layers = [],
    ...rootRest
  } = props;

  function renderLayers() {
    if (targetElement) {
      const shouldUseLayers = layers && layers.length > 0;
      if (shouldUseLayers) {
        return layers.map((layer, i) => (
          <ParallaxBannerLayer
            {...layer}
            targetElement={targetElement}
            key={`layer-${i}`}
            testId={`layer-${i}`}
          />
        ));
      }
    }
    return null;
  }

  function renderChildren() {
    if (targetElement) {
      return React.Children.map(props.children, (child) => {
        const item = child as ReactElement<
          PropsWithChildren<{ targetElement: any }>
        >;
        // adds the targetElement prop to any ParallaxBannerLayer components
        if (item?.type === ParallaxBannerLayer) {
          const clone = React.cloneElement(item, {
            targetElement,
          });
          return clone;
        }
        return child;
      });
    }
    return null;
  }
  return (
    <div
      ref={containerRef}
      style={{ ...containerStyle, ...rootStyle }}
      {...rootRest}
    >
      {/* Using the `layers` prop to define children */}
      {renderLayers()}
      {/* Using children to compose layers */}
      {renderChildren()}
    </div>
  );
};
