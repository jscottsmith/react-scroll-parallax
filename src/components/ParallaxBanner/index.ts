export { ParallaxBanner } from './ParallaxBanner';
export { ParallaxBannerLayer } from './components/ParallaxBannerLayer';
export * from './types';
