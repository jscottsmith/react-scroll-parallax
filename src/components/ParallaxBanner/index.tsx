import React, {
  PropsWithChildren,
  CSSProperties,
  useEffect,
  useRef,
  useState,
} from 'react';
import { ParallaxBannerLayer } from './components/ParallaxBannerLayer';
import { ParallaxBannerProps } from './types';

const containerStyle: CSSProperties = {
  position: 'relative',
  overflow: 'hidden',
  width: '100%',
};

export const ParallaxBanner = (
  props: PropsWithChildren<ParallaxBannerProps>
) => {
  const [targetElement, setTargetElement] = useState<HTMLDivElement | null>(
    null
  );
  const containerRef = useRef<HTMLDivElement>(null);
  useEffect(() => {
    setTargetElement(containerRef.current);
  }, []);
  const {
    disabled: disableAllLayers,
    style: rootStyle,
    layers = [],
    ...rootRest
  } = props;
  return (
    <div
      ref={containerRef}
      style={{ ...containerStyle, ...rootStyle }}
      {...rootRest}
    >
      {layers.map(
        (layer, i) =>
          targetElement && (
            <ParallaxBannerLayer
              {...layer}
              targetElement={targetElement}
              key={`layer-${i}`}
              testId={`layer-${i}`}
            />
          )
      )}
      {props.children}
    </div>
  );
};
