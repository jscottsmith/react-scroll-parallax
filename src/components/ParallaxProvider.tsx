import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ParallaxContext from '../helpers/ParallaxContext';
import ParallaxController from '../classes/ParallaxController';
import { VERTICAL, HORIZONTAL } from '../constants';
import validHTMLElement from '../utils/validHTMLElement';

const createController = (options) => {
  // Don't initialize on the server
  const isServer = typeof window === 'undefined';

  if (!isServer) {
    // Must not be the server so kick it off...
    return ParallaxController.init(options);
  }
  return null;
};

export interface ParallaxProviderProps {
  /**
   * Optionally pass the scroll axis for setting horizontal/vertical scrolling. One of vertical or
   * horizontal
   */
  scrollAxis?: 'vertical' | 'horizontal';
  /**
   * Optionally set the container that has overflow and will contain parallax elements. Defaults
   * to the HTML body
   */
  scrollContainer?: any;
}

export default class ParallaxProvider extends Component<
  ParallaxProviderProps,
  {}
> {
  static defaultProps = {
    scrollAxis: VERTICAL,
  };

  static propTypes = {
    children: PropTypes.node.isRequired,
    scrollAxis: PropTypes.oneOf([VERTICAL, HORIZONTAL]),
    scrollContainer: validHTMLElement,
  };

  constructor(props) {
    super(props);
    // @ts-ignore
    this.controller = createController({
      scrollAxis: props.scrollAxis,
      scrollContainer: props.scrollContainer,
    });
  }

  componentDidUpdate(prevProps) {
    if (prevProps.scrollContainer !== this.props.scrollContainer) {
      // @ts-ignore
      this.controller.updateScrollContainer(this.props.scrollContainer);
    }
  }

  componentWillUnmount() {
    // @ts-ignore
    this.controller = this.controller.destroy();
  }

  render() {
    const { children } = this.props;
    return (
      // @ts-ignore
      <ParallaxContext.Provider value={this.controller}>
        {children}
      </ParallaxContext.Provider>
    );
  }
}
