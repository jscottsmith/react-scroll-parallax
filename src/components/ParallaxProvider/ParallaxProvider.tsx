import React, { Component } from 'react';

import { ParallaxContext } from '../../context/ParallaxContext';
import { ParallaxController, ScrollAxis } from 'parallax-controller';
import { ParallaxProviderProps } from './types';
import { createController } from './helpers';

export class ParallaxProvider extends Component<ParallaxProviderProps, {}> {
  static defaultProps = {
    scrollAxis: ScrollAxis.vertical,
  };

  controller: ParallaxController | null;

  constructor(props: ParallaxProviderProps) {
    super(props);
    this.controller = createController({
      scrollAxis: props.scrollAxis,
      scrollContainer: props.scrollContainer,
      disabled: props.isDisabled,
    });
  }

  componentDidUpdate(prevProps: ParallaxProviderProps) {
    if (
      prevProps.scrollContainer !== this.props.scrollContainer &&
      this.props.scrollContainer
    ) {
      this.controller?.updateScrollContainer(this.props.scrollContainer);
    }

    if (prevProps.isDisabled !== this.props.isDisabled) {
      if (this.props.isDisabled) {
        this.controller?.disableParallaxController();
      }
      if (!this.props.isDisabled) {
        this.controller?.enableParallaxController();
      }
    }
  }

  componentWillUnmount() {
    // @ts-ignore
    this.controller = this.controller.destroy();
  }

  render() {
    const { children } = this.props;
    return (
      // @ts-ignore
      <ParallaxContext.Provider value={this.controller}>
        {children}
      </ParallaxContext.Provider>
    );
  }
}
