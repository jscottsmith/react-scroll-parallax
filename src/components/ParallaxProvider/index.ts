export { ParallaxProvider } from './ParallaxProvider';
export * from './types';
