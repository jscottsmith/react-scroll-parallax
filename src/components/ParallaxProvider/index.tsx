import React, { Component } from 'react';

import { ParallaxContext } from '../../context/ParallaxContext';
import { ParallaxController, ScrollAxis } from 'parallax-controller';
import { ParallaxProviderProps } from './types';
import { createController } from './helpers';

export class ParallaxProvider extends Component<ParallaxProviderProps, {}> {
  static defaultProps = {
    scrollAxis: ScrollAxis.vertical,
  };

  controller: ParallaxController | null;

  constructor(props: ParallaxProviderProps) {
    super(props);
    this.controller = createController({
      scrollAxis: props.scrollAxis,
      scrollContainer: props.scrollContainer,
    });
  }

  componentDidUpdate(prevProps: ParallaxProviderProps) {
    if (prevProps.scrollContainer !== this.props.scrollContainer) {
      // @ts-ignore
      this.controller.updateScrollContainer(this.props.scrollContainer);
    }
  }

  componentWillUnmount() {
    // @ts-ignore
    this.controller = this.controller.destroy();
  }

  render() {
    const { children } = this.props;
    return (
      // @ts-ignore
      <ParallaxContext.Provider value={this.controller}>
        {children}
      </ParallaxContext.Provider>
    );
  }
}
