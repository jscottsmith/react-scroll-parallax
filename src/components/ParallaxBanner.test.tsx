import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import createNodeMock from '../testUtils/createNodeMock';
import { ParallaxBanner } from './ParallaxBanner';
import { ParallaxProvider } from './ParallaxProvider';

describe('Expect the <ParallaxBanner> component', () => {
  afterEach(() => {});

  it('to render image banners correctly', () => {
    const tree = renderer
      .create(
        <ParallaxProvider>
          <ParallaxBanner
            className="test-class"
            disabled={false}
            layers={[
              {
                image: 'https://foo.com/bar.jpg',
                amount: 0.2,
              },
            ]}
            style={{
              backgroundColor: 'blue',
              border: '1px solid red',
            }}
          >
            <div>
              <h1>Foo Bar</h1>
            </div>
          </ParallaxBanner>
        </ParallaxProvider>,
        {
          createNodeMock,
        }
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('to render custom child banners correctly', () => {
    const tree = renderer
      .create(
        <ParallaxProvider>
          <ParallaxBanner
            className="test-class"
            disabled={false}
            layers={[
              {
                children: <div>test</div>,
                amount: 0.2,
              },
            ]}
            style={{
              backgroundColor: 'blue',
              border: '1px solid red',
            }}
          />
        </ParallaxProvider>,
        {
          createNodeMock,
        }
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('to render without expanded margins', () => {
    const tree = renderer
      .create(
        <ParallaxProvider>
          <ParallaxBanner
            layers={[
              {
                children: <div>test</div>,
                amount: 0.2,
                expanded: false,
              },
            ]}
          />
        </ParallaxProvider>,
        {
          createNodeMock,
        }
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('to render children', () => {
    const node = document.createElement('div');

    const child = jest.fn();
    const Child = () => {
      child();
      return <div />;
    };

    ReactDOM.render(
      <ParallaxProvider>
        <ParallaxBanner layers={[]}>
          <Child />
        </ParallaxBanner>
      </ParallaxProvider>,
      node
    );

    expect(child).toBeCalled();
  });

  it('to render layer children', () => {
    const node = document.createElement('div');

    const childFn = jest.fn();
    const Child = () => {
      childFn();
      return <div />;
    };

    ReactDOM.render(
      <ParallaxProvider>
        <ParallaxBanner
          layers={[
            {
              children: <Child />,
              amount: 0.2,
            },
          ]}
        />
      </ParallaxProvider>,
      node
    );

    expect(childFn).toBeCalled();
  });

  it('to render layers with custom props', () => {
    const tree = renderer
      .create(
        <ParallaxProvider>
          <ParallaxBanner
            layers={[
              {
                amount: 0.2,
                props: {
                  style: {
                    backgroundColor: 'red',
                  },
                  className: 'my-custom-class',
                  id: 'my-id',
                },
              },
            ]}
          />
        </ParallaxProvider>,
        {
          createNodeMock,
        }
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
