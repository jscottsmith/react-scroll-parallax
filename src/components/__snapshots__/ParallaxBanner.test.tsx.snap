// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect the <ParallaxBanner> component to render custom child banners correctly 1`] = `
<div
  className="parallax-banner test-class"
  style={
    Object {
      "backgroundColor": "blue",
      "border": "1px solid red",
      "height": "50vh",
      "overflow": "hidden",
      "position": "relative",
      "width": "100%",
    }
  }
>
  <div
    className="parallax-outer"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <div
      className="parallax-inner"
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <div
        className="parallax-banner-layer-0"
        style={
          Object {
            "bottom": "-20%",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": "-20%",
          }
        }
      >
        <div>
          test
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Expect the <ParallaxBanner> component to render image banners correctly 1`] = `
<div
  className="parallax-banner test-class"
  style={
    Object {
      "backgroundColor": "blue",
      "border": "1px solid red",
      "height": "50vh",
      "overflow": "hidden",
      "position": "relative",
      "width": "100%",
    }
  }
>
  <div
    className="parallax-outer"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <div
      className="parallax-inner"
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <div
        className="parallax-banner-layer-0"
        style={
          Object {
            "backgroundImage": "url(https://foo.com/bar.jpg)",
            "backgroundPosition": "center",
            "backgroundSize": "cover",
            "bottom": "-20%",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": "-20%",
          }
        }
      />
    </div>
  </div>
  <div>
    <h1>
      Foo Bar
    </h1>
  </div>
</div>
`;

exports[`Expect the <ParallaxBanner> component to render layers with custom props 1`] = `
<div
  className="parallax-banner"
  style={
    Object {
      "height": "50vh",
      "overflow": "hidden",
      "position": "relative",
      "width": "100%",
    }
  }
>
  <div
    className="parallax-outer"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <div
      className="parallax-inner"
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <div
        className="parallax-banner-layer-0 my-custom-class"
        id="my-id"
        style={
          Object {
            "backgroundColor": "red",
            "bottom": "-20%",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": "-20%",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`Expect the <ParallaxBanner> component to render without expanded margins 1`] = `
<div
  className="parallax-banner"
  style={
    Object {
      "height": "50vh",
      "overflow": "hidden",
      "position": "relative",
      "width": "100%",
    }
  }
>
  <div
    className="parallax-outer"
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <div
      className="parallax-inner"
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <div
        className="parallax-banner-layer-0"
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <div>
          test
        </div>
      </div>
    </div>
  </div>
</div>
`;
